### PR TITLE
support network neighbors

### PR DESCRIPTION
**Pull request by Daniel-GrunbergerCA**
## PR Type:
Enhancement

___
## PR Description:
This PR introduces a new feature to support network neighbors. The main changes include:
- Adding a network tracer to listen to communication.
- Adding a network manager to handle network events.
- Supporting services without selectors.
- Adding tests and debug logs.
- Fixing label selector.
- Handling network events and generating entries for CRD.
- Creating and updating network neighbors CRD based on events.
- Handling services with no selectors.
- Deleting resources when a container is removed.

___
## PR Main Files Walkthrough:
<details> <summary>files:</summary>

`pkg/networkmanager/network_manager.go`: This file introduces the NetworkManager struct and its associated methods. The NetworkManager is responsible for handling network events, creating and updating network neighbors CRD based on these events, and deleting resources when a container is removed.
`pkg/containerwatcher/v1/open_test.go`: This file modifies the test setup to accommodate the new NetworkManager parameter in the CreateIGContainerWatcher function.
`pkg/networkmanager/network_manager_test.go`: This file includes tests for the NetworkManager, ensuring its methods function as expected.
`pkg/networkmanager/network_neighbors.go`: This file introduces functions to generate network neighbors CRD and network neighbors entries.
`pkg/containerwatcher/v1/container_watcher.go`: This file modifies the ContainerWatcher struct to include a NetworkManager, and updates the relevant methods to use this manager.
`pkg/networkmanager/network_event.go`: This file introduces the NetworkEvent struct and its associated methods. NetworkEvent represents a network event that occurs in a container.
`pkg/storage/storage_interface.go`: This file updates the StorageClient interface to include methods for creating, getting, patching, and deleting NetworkNeighbors.
`pkg/relevancymanager/v1/relevancy_manager.go`: This file modifies the RelevancyManager to include a NetworkManager, and updates the relevant methods to use this manager.
`pkg/config/config.go`: This file updates the Config struct to include parameters related to the NetworkManager, such as InitialDelay and UpdateDataPeriod.
`main.go`: This file modifies the main function to create a NetworkManager and pass it to the RelevancyManager and ContainerWatcher.
</details>

___
## User Description:
1. Add a network tracer to listen to communication.
2. Add a network manager to handle network events

### CODE REVIEW

File: pkg/networkmanager/network_event.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+	ne.Destination.PodLabels = generatePodLabels(podLabels)
+}
+
+func generatePodLabels(podLabels map[string]string) string {
+	var keys []string
+	for key := range podLabels {
+		keys = append(keys, key)
+	}
+
+	sort.Strings(keys)
```
Comment (matthyx) [inline pkg/networkmanager/network_event.go:75]: ;-)

**Review by matthyx — COMMENTED** [review]
integration in the node-agent LGTM, unfortunately I don't have enough time to validate the business logic of generating network-policies

File: go.mod [hunk 2]
```diff
@@ -7,21 +7,24 @@ require (
...
+	github.com/google/uuid v1.3.1
 	github.com/goradd/maps v0.1.5
-	github.com/inspektor-gadget/inspektor-gadget v0.20.1-0.20230919195508-052fbf78b54c
+	github.com/inspektor-gadget/inspektor-gadget v0.21.0
 	github.com/kinbiko/jsonassert v1.1.1
 	github.com/kubescape/go-logger v0.0.21
-	github.com/kubescape/k8s-interface v0.0.145
-	github.com/kubescape/storage v0.0.22
+	github.com/kubescape/k8s-interface v0.0.148
+	github.com/kubescape/storage v0.0.30-0.20231024145442-963c4296f5b4
```
Comment (dwertent) [inline go.mod:16]: Can we release the storage?

Comment (Daniel-GrunbergerCA) [reply go.mod:16]: waiting for vlad approval

File: pkg/storage/v1/storage_nocache.go [hunk 3]
```diff
@@ -30,7 +29,7 @@ var _ storage.StorageClient = (*StorageNoCache)(nil)
 
 func CreateStorageNoCache(namespace string) (*StorageNoCache, error) {
 	var config *rest.Config
-	kubeconfig := os.Getenv(KubeConfig)
+	kubeconfig := "/home/daniel/.kube/config"
```
Comment (dwertent) [inline pkg/storage/v1/storage_nocache.go:32]:
```suggestion
	kubeconfig := "/home/daniel/.kube/config"
```

File: pkg/relevancymanager/v1/relevancy_manager.go [hunk 4]
```diff
@@ -150,7 +150,7 @@ func (rm *RelevancyManager) handleRelevancy(ctx context.Context, watchedContaine
 		logger.L().Debug("failed to get file list", helpers.String("container ID", containerID), helpers.String("k8s workload", watchedContainer.K8sContainerID), helpers.Error(err))
 		return
 	}
-	logger.L().Debug("fileList generated", helpers.String("container ID", containerID), helpers.String("k8s workload", watchedContainer.K8sContainerID), helpers.String("file list", fmt.Sprintf("%v", fileList)))
+	// logger.L().Debug("fileList generated", helpers.String("container ID", containerID), helpers.String("k8s workload", watchedContainer.K8sContainerID), helpers.String("file list", fmt.Sprintf("%v", fileList)))
```
Comment (dwertent) [inline pkg/relevancymanager/v1/relevancy_manager.go:153]: Remove

File: pkg/networkmanager/network_manager.go [hunk 5]
```diff
@@ -0,0 +1,543 @@
...
+	parentWL, err := am.getParentWorkloadFromContainer(container)
+	if err != nil {
+		logger.L().Info("NetworkManager - failed to get parent workload", helpers.String("reason", err.Error()), helpers.String("container ID", container.Runtime.ContainerID), helpers.String("k8s workload", k8sContainerID))
+		return
+	}
+
+	selector, err := parentWL.GetSelector()
+	if err != nil {
+		// if we get not selector, we can't create/update network neighbor
+		logger.L().Info("NetworkManager - failed to get selector", helpers.String("reason", err.Error()), helpers.String("container ID", container.Runtime.ContainerID), helpers.String("k8s workload", k8sContainerID))
```
Comment (dwertent) [inline pkg/networkmanager/network_manager.go:164]: Change log level

File: pkg/networkmanager/network_manager.go [hunk 6]
```diff
@@ -0,0 +1,543 @@
...
+		return
+	}
+
+	if selector == nil {
+		// if we get not selector, we can't create/update network neighbor
+		logger.L().Info("NetworkManager - selector is nil", helpers.String("container ID", container.Runtime.ContainerID), helpers.String("k8s workload", k8sContainerID))
```
Comment (dwertent) [inline pkg/networkmanager/network_manager.go:170]: Change log level

File: pkg/networkmanager/network_manager.go [hunk 7]
```diff
@@ -0,0 +1,543 @@
...
+		return
+	}
+
+	// check if network neighbor CRD exists
+	networkNeighbors, err := am.storageClient.GetNetworkNeighbors(parentWL.GetNamespace(), generateNetworkNeighborsNameFromWorkload(parentWL))
+	if err != nil {
+		if !strings.Contains(err.Error(), "not found") {
+			logger.L().Info("NetworkManager - failed to get network neighbor", helpers.String("reason", err.Error()), helpers.String("container ID", container.Runtime.ContainerID), helpers.String("k8s workload", k8sContainerID))
```
Comment (dwertent) [inline pkg/networkmanager/network_manager.go:178]: Change log level

File: pkg/networkmanager/network_manager.go [hunk 8]
```diff
@@ -0,0 +1,543 @@
...
+		} else {
+			// network neighbor not found, create new one
+			newNetworkNeighbors := generateNetworkNeighborsCRD(parentWL, selector)
+
+			if err = am.storageClient.CreateNetworkNeighbors(newNetworkNeighbors, parentWL.GetNamespace()); err != nil {
+				logger.L().Info("NetworkManager - failed to create network neighbor", helpers.String("reason", err.Error()), helpers.String("container ID", container.Runtime.ContainerID), helpers.String("k8s workload", k8sContainerID))
```
Comment (dwertent) [inline pkg/networkmanager/network_manager.go:184]: Change log level

**Review by dwertent — CHANGES_REQUESTED** [review]

**Review by dwertent — APPROVED** [review]

**Pull request merged**